File: Python/ByRound/2043/2043_E_Matrix_Transformation.py
# Problem: CF 2043 E - Matrix Transformation
# https://codeforces.com/contest/2043/problem/E

"""
Code Purpose:
This code determines whether it's possible to transform matrix A into matrix B using two types of operations:
1. Bitwise AND operation on entire rows
2. Bitwise OR operation on entire columns

The solution works by analyzing each bit position independently:
- For each bit position, it checks if the transformation is possible by simulating
  the constraints imposed by the operations on rows and columns.
- It uses a greedy approach to determine which rows and columns can be "disabled"
  (set to 0 or 1 respectively) through operations, then verifies that all remaining
  active cells match the target.

Algorithms/Techniques:
- Bit Manipulation
- Graph-based greedy approach with row/column constraint simulation
- Bit-by-bit checking (20 bits for 10^9 values)

Time Complexity: O(20 * n * m) where 20 is the number of bits (max 30 bits for 10^9)
Space Complexity: O(n + m) for storing counts and active flags
"""

import sys

input = sys.stdin.readline
K = 20


def solve():
    n, m = map(int, input().split())
    A = [list(map(int, input().split())) for _ in range(n)]
    B = [list(map(int, input().split())) for _ in range(n)]

    def possible(bit):
        # Count how many elements in each row/column will become 0 or 1 after operations
        row_zero_cnt = [0] * n  # how many 0s in each row
        col_one_cnt = [0] * m   # how many 1s in each column

        # Count the required transitions for each bit
        for i in range(n):
            for j in range(m):
                if B[i][j] >> bit & 1:  # if bit at position `bit` is 1 in B
                    col_one_cnt[j] += 1
                else:  # if bit at position `bit` is 0 in B
                    row_zero_cnt[i] += 1

        # Group rows and columns by their constraints
        col_zeros = [[] for _ in range(n + 1)]  # rows where bit is 0 in B, grouped by how many 0s
        row_ones = [[] for _ in range(m + 1)]   # cols where bit is 1 in B, grouped by how many 1s

        # Fill groups with indices that meet the criteria for that row/column
        for i, cnt in enumerate(row_zero_cnt):
            row_ones[m - cnt].append(i)

        for i, cnt in enumerate(col_one_cnt):
            col_zeros[n - cnt].append(i)

        # Set up active row/column flags
        active_rows = [True] * n
        active_cols = [True] * m
        deleted_rows = 0
        deleted_cols = 0

        # Greedily eliminate rows/columns where we can't change the result
        while col_zeros[deleted_rows] or row_ones[deleted_cols]:
            # Process columns that must stay 0 (due to constraints)
            while col_zeros[deleted_rows]:
                deleted_cols += 1
                active_cols[col_zeros[deleted_rows].pop()] = False

            # Process rows that must stay 1 (due to constraints)
            while row_ones[deleted_cols]:
                deleted_rows += 1
                active_rows[row_ones[deleted_cols].pop()] = False

        # Check if any remaining cell doesn't match target
        for i in range(n):
            for j in range(m):
                if (
                    active_cols[j]
                    and active_rows[i]
                    and A[i][j] >> bit & 1 != B[i][j] >> bit & 1
                ):
                    return False

        return True

    # Check all 20 bits to ensure transformation is possible
    for b in range(K):
        if not possible(b):
            return "No"
    return "Yes"


T = int(input())
out = [solve() for _ in range(T)]
print("\n".join(map(str, out)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/2043_E_Matrix_Transformation.cpp
// Problem: CF 2043 E - Matrix Transformation
// https://codeforces.com/contest/2043/problem/E

/*
E. Matrix Transformation

Purpose:
This code determines whether it is possible to transform matrix A into matrix B using two types of operations:
- Bitwise AND operation on a row (A[i][j] &= x)
- Bitwise OR operation on a column (A[i][j] |= x)

The approach processes the matrices bit by bit, checking for each bit position whether it's possible to achieve the target values in B from A using these operations.
It uses a reduction to a graph problem where each row/column is treated as a node and constraints between them define edges. Using a topological sort approach with DFS to detect cycles, we verify consistency across bit positions.

Algorithms/Techniques:
- Bit manipulation
- Graph theory (topological ordering using DFS)
- Binary search on bits (process one bit at a time)

Time Complexity: O(30 * n * m), where 30 is the number of bits for integers.
Space Complexity: O(n + m), due to graph representation and auxiliary vectors.

*/

#include <limits.h>
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
#define lc f << 1
#define rc f << 1 | 1
#define ll long long
#define pii pair<int, int>
#define fi first
#define se second
#define lowbit(x) x & -x
#define gcd(x, y) __gcd(abs(x), abs(y))
const ll inf = LONG_LONG_MAX, mod = 998244353;
const int N = 1e7 + 10, inff = INT_MAX;
const double eps = 1e-9;

void solve() {
  int n, m;
  cin >> n >> m;
  vector<vector<int>> x(n + 1, vector<int>(m + 1, 0)), // matrix A
      y(n + 1, vector<int>(m + 1, 0)); // matrix B
  vector<vector<int>> a(n + 1, vector<int>(m + 1, 0)),
      b(n + 1, vector<int>(m + 1, 0)); // bit representations of A and B
  for (int i = 1; i <= n; i++)
    for (int j = 1; j <= m; j++)
      cin >> x[i][j]; // input matrix A
  for (int i = 1; i <= n; i++)
    for (int j = 1; j <= m; j++)
      cin >> y[i][j]; // input matrix B
  bool f = 1; // flag to continue processing bits

  auto check = [&]() -> bool {
    vector<vector<int>> e(n + m + 1); // graph edges (n rows + m columns)
    vector<bool> v(n + m + 1, 0), in(n + m + 1, 0), vis(n + m + 1, 0); // v: marks if node is a constraint node; in: dfs recursion stack; vis: visited flag
    auto dfs = [&](int s, auto &dfs) -> bool {
      if (in[s]) // cycle detected
        return 0;
      in[s] = 1; // mark as being in DFS traversal
      for (auto it : e[s])
        if (!vis[it] && !dfs(it, dfs)) 
          return 0;
      in[s] = 0; // remove from recursion stack
      vis[s] = 1; // mark visited
      return 1;
    };
    for (int i = 1; i <= n; i++)
      for (int j = 1; j <= m; j++) {
        if (b[i][j]) // if b[i][j] is 1, then it requires that a[i][j] must be at least 1 after some operation
          e[i].push_back(j + n); // add constraint from row node i to column node j+n
        else
          e[j + n].push_back(i); // add reverse constraint if b[i][j] is 0, then a[i][j] should be 0 or less
        if (a[i][j] && !b[i][j]) // if bit in A is 1 but B has 0, this means we must prevent that
          v[i] = 1; // mark row as constrained
        if (!a[i][j] && b[i][j]) // if A is 0 but B has 1, column constraint needed
          v[j + n] = 1; // mark column as constrained
      }
    for (int i = 1; i <= n + m; i++) // check constraints
      if (v[i] && !dfs(i, dfs)) // if a node is marked as violated and not acyclic
        return 0;
    return 1;
  };

  while (f) {
    f = 0;
    for (int i = 1; i <= n; i++)
      for (int j = 1; j <= m; j++) {
        a[i][j] = x[i][j] % 2; // extract bit from A
        x[i][j] /= 2; // shift right to process next bit
        if (x[i][j])
          f = 1;
      }
    for (int i = 1; i <= n; i++)
      for (int j = 1; j <= m; j++) {
        b[i][j] = y[i][j] % 2; // extract bit from B
        y[i][j] /= 2; // shift right to process next bit
        if (y[i][j])
          f = 1;
      }
    if (!check()) { // check consistency of current bit level
      cout << "No\n";
      return;
    }
  }
  cout << "Yes\n"; // all bits are consistent, transformation possible
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(0), cout.tie(0);
  int T = 1;
  cin >> T;
  while (T--)
    solve();
  return 0;
}


// https://github.com/VaHiX/codeForces/